File: frontend/src/types/candidate.ts
export type PipelineStage = "sourced" | "screening" | "interview" | "submitted" | "placed";
export type ApplicationStatus = "active" | "on_hold" | "rejected" | "withdrawn" | "placed";
export type ParseStatus = "pending" | "processing" | "completed" | "failed";
export type ResumeVersionStatus = "draft" | "pending_review" | "approved" | "needs_changes" | "rejected";
export type RecruiterReviewDecision =
  | "accept"
  | "reject"
  | "needs_more_interview"
  | "needs_resume_changes"
  | "submit_to_client";
export type InterviewStatus = "scheduled" | "in_progress" | "completed" | "cancelled";
export type HireRecommendation = "strong_hire" | "hire" | "hold" | "reject";

export interface Candidate {
  id: string;
  first_name: string;
  last_name: string;
  full_name: string;
  email: string | null;
  phone: string | null;
  linkedin_url: string | null;
  github_url: string | null;
  portfolio_url: string | null;
  current_company: string | null;
  current_ctc: string | null;
  expected_ctc: string | null;
  notice_period_days: number | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface CandidateListResponse {
  items: Candidate[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface CandidateCreatePayload {
  first_name: string;
  last_name: string;
  email?: string;
  phone?: string;
  linkedin_url?: string;
  github_url?: string;
  portfolio_url?: string;
  current_company?: string;
  current_ctc?: number;
  expected_ctc?: number;
  notice_period_days?: number;
  notes?: string;
  job_requirement_id?: string;
}

export interface ParsedResume {
  id: string;
  application_id: string;
  status: ParseStatus;
  structured_data: Record<string, unknown> | null;
  error_message: string | null;
  parsed_at: string | null;
}

export interface ResumeVersion {
  id: string;
  application_id: string;
  version_number: number;
  status: ResumeVersionStatus;
  template_id: string | null;
  template_name?: string | null;
  content_json: Record<string, unknown>;
  recruiter_review_decision: RecruiterReviewDecision | null;
  recruiter_review_notes: string | null;
  reviewed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface ResumeScore {
  id: string;
  application_id: string;
  resume_version_id: string | null;
  overall_score: string;
  keyword_match: string | null;
  skill_match: string | null;
  experience_match: string | null;
  semantic_similarity: string | null;
  formatting_score: string | null;
  grammar_score: string | null;
  achievements_score: string | null;
  missing_keywords: string[];
  suggestions: string[];
  improvement_areas: string[];
  created_at: string;
}

export interface ScreeningInterview {
  id: string;
  application_id: string;
  status: InterviewStatus;
  scheduled_at: string | null;
  duration_minutes: number;
  difficulty: string;
  coding_required: boolean;
  behavioral: boolean;
  technical: boolean;
  language: string;
  timezone: string;
  interview_link: string | null;
  transcript: Array<Record<string, unknown>>;
  summary: string | null;
  technical_score: string | null;
  coding_score: string | null;
  communication_score: string | null;
  confidence_score: string | null;
  problem_solving_score: string | null;
  recommendation: HireRecommendation | null;
  completed_at: string | null;
}

export interface ApplicationSummary {
  id: string;
  candidate_id: string;
  candidate_name: string;
  job_requirement_id: string;
  job_title: string;
  client_name: string;
  pipeline_stage: PipelineStage;
  status: ApplicationStatus;
  has_parsed_resume: boolean;
  resume_version_count: number;
  latest_resume_version_id: string | null;
  latest_built_resume_status: ResumeVersionStatus | null;
  latest_score: string | null;
  interview_status: InterviewStatus | null;
  interview_overall_score: string | null;
  created_at: string;
  updated_at: string;
}

export interface ApplicationDetail {
  id: string;
  tenant_id: string;
  candidate_id: string;
  candidate: Candidate;
  job_requirement_id: string;
  job_title: string;
  client_name: string;
  job_description: string | null;
  job_experience_min_years: number | null;
  job_experience_max_years: number | null;
  pipeline_stage: PipelineStage;
  status: ApplicationStatus;
  recruiter_notes: string | null;
  parsed_resume: ParsedResume | null;
  resume_versions: ResumeVersion[];
  resume_scores: ResumeScore[];
  screening_interview: ScreeningInterview | null;
  created_at: string;
  updated_at: string;
}

export interface ApplicationListResponse {
  items: ApplicationSummary[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface ResumeExperienceInput {
  company: string;
  title: string;
  start_date: string;
  end_date?: string | null;
  is_current?: boolean;
  bullets?: string[];
}

export interface ResumeEducationInput {
  institution: string;
  degree: string;
  field?: string | null;
  graduation_year?: string | null;
  start_year?: string | null;
}

export interface TimelineGap {
  label: string;
  from_date: string;
  to_date: string;
  months: number;
  suggestion: string;
}

export interface GapStrategyOption {
  id: string;
  label: string;
  description: string;
}

export interface ResumeBuildContext {
  job_experience_min_years: number | null;
  job_experience_max_years: number | null;
  candidate_total_experience_years: number;
  experience_shortfall_years: number | null;
  timeline_gaps: TimelineGap[];
  gap_strategy_options: GapStrategyOption[];
  recommendations: string[];
  default_summary: string | null;
  default_skills: string[];
  default_experience: ResumeExperienceInput[];
  default_education: ResumeEducationInput[];
  templates: Array<{
    id: string;
    name: string;
    description: string | null;
    is_default: boolean;
    is_system: boolean;
    source_type: string;
    source_file_name: string | null;
  }>;
  default_template_id: string | null;
}

export interface BuildResumePayload {
  template_id: string;
  recruiter_notes?: string;
  gap_strategy?: string;
  target_total_experience_years?: number;
  summary?: string;
  skills?: string[];
  experience?: ResumeExperienceInput[];
  education?: ResumeEducationInput[];
  training_entry?: ResumeExperienceInput;
}

export interface InterviewQuestionResponse {
  question: string;
  interview_status: InterviewStatus;
  question_number: number;
  question_type: "mcq" | "scenario" | "coding";
  options: Array<{ id: string; text: string }>;
  phase_label?: string | null;
  total_questions: number;
  is_complete: boolean;
}

export const PIPELINE_STAGE_LABELS: Record<PipelineStage, string> = {
  sourced: "Sourced",
  screening: "Screening",
  interview: "Interview",
  submitted: "Submitted",
  placed: "Placed",
};

export const PIPELINE_STEPS = [
  { key: "create", label: "Create Candidate" },
  { key: "upload", label: "Upload Resume" },
  { key: "parse", label: "AI Parse" },
  { key: "build", label: "AI Resume Builder" },
  { key: "review", label: "Recruiter Review" },
  { key: "export", label: "Export PDF/DOCX" },
  { key: "score", label: "ATS Score" },
  { key: "schedule", label: "Schedule Interview" },
  { key: "interview", label: "AI Interview" },
  { key: "summary", label: "Interview Summary" },
] as const;

File: backend/app/schemas/candidate.py
"""Pydantic schemas for candidate management and pipeline."""

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.schemas.resume_template import ResumeTemplateSummary


class PipelineStage(str, Enum):
    SOURCED = "sourced"
    SCREENING = "screening"
    INTERVIEW = "interview"
    SUBMITTED = "submitted"
    PLACED = "placed"


class ApplicationStatus(str, Enum):
    ACTIVE = "active"
    ON_HOLD = "on_hold"
    REJECTED = "rejected"
    WITHDRAWN = "withdrawn"
    PLACED = "placed"


class ParseStatus(str, Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class ResumeVersionStatus(str, Enum):
    DRAFT = "draft"
    PENDING_REVIEW = "pending_review"
    APPROVED = "approved"
    NEEDS_CHANGES = "needs_changes"
    REJECTED = "rejected"


class RecruiterReviewDecision(str, Enum):
    ACCEPT = "accept"
    REJECT = "reject"
    NEEDS_MORE_INTERVIEW = "needs_more_interview"
    NEEDS_RESUME_CHANGES = "needs_resume_changes"
    SUBMIT_TO_CLIENT = "submit_to_client"


class InterviewStatus(str, Enum):
    SCHEDULED = "scheduled"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class HireRecommendation(str, Enum):
    STRONG_HIRE = "strong_hire"
    HIRE = "hire"
    HOLD = "hold"
    REJECT = "reject"


class CandidateCreate(BaseModel):
    first_name: str = Field(..., min_length=1, max_length=100)
    last_name: str = Field(..., min_length=1, max_length=100)
    email: EmailStr | None = None
    phone: str | None = Field(default=None, max_length=50)
    linkedin_url: str | None = Field(default=None, max_length=512)
    github_url: str | None = Field(default=None, max_length=512)
    portfolio_url: str | None = Field(default=None, max_length=512)
    current_company: str | None = Field(default=None, max_length=255)
    current_ctc: Decimal | None = Field(default=None, ge=0)
    expected_ctc: Decimal | None = Field(default=None, ge=0)
    notice_period_days: int | None = Field(default=None, ge=0, le=365)
    notes: str | None = Field(default=None, max_length=5000)
    job_requirement_id: UUID | None = None


class CandidateUpdate(BaseModel):
    first_name: str | None = Field(default=None, min_length=1, max_length=100)
    last_name: str | None = Field(default=None, min_length=1, max_length=100)
    email: EmailStr | None = None
    phone: str | None = None
    linkedin_url: str | None = None
    github_url: str | None = None
    portfolio_url: str | None = None
    current_company: str | None = None
    current_ctc: Decimal | None = None
    expected_ctc: Decimal | None = None
    notice_period_days: int | None = None
    notes: str | None = None


class CandidateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    first_name: str
    last_name: str
    full_name: str
    email: str | None
    phone: str | None
    linkedin_url: str | None
    github_url: str | None
    portfolio_url: str | None
    current_company: str | None
    current_ctc: Decimal | None
    expected_ctc: Decimal | None
    notice_period_days: int | None
    notes: str | None
    created_at: datetime
    updated_at: datetime


class CandidateListResponse(BaseModel):
    items: list[CandidateResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class ApplicationCreate(BaseModel):
    candidate_id: UUID
    job_requirement_id: UUID
    recruiter_notes: str | None = None


class ParsedResumeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    application_id: UUID
    status: ParseStatus
    structured_data: dict[str, Any] | None
    error_message: str | None
    parsed_at: datetime | None


class ResumeVersionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    application_id: UUID
    version_number: int
    status: ResumeVersionStatus
    template_id: UUID | None = None
    template_name: str | None = None
    content_json: dict[str, Any]
    recruiter_review_decision: RecruiterReviewDecision | None
    recruiter_review_notes: str | None
    reviewed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class ResumeScoreResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    application_id: UUID
    resume_version_id: UUID | None
    overall_score: Decimal
    keyword_match: Decimal | None
    skill_match: Decimal | None
    experience_match: Decimal | None
    semantic_similarity: Decimal | None
    formatting_score: Decimal | None
    grammar_score: Decimal | None
    achievements_score: Decimal | None
    missing_keywords: list[str]
    suggestions: list[str]
    improvement_areas: list[str]
    created_at: datetime


class ScreeningInterviewResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    application_id: UUID
    status: InterviewStatus
    scheduled_at: datetime | None
    duration_minutes: int
    difficulty: str
    coding_required: bool
    behavioral: bool
    technical: bool
    language: str
    timezone: str
    interview_link: str | None
    transcript: list[dict[str, Any]]
    summary: str | None
    technical_score: Decimal | None
    coding_score: Decimal | None
    communication_score: Decimal | None
    confidence_score: Decimal | None
    problem_solving_score: Decimal | None
    recommendation: HireRecommendation | None
    completed_at: datetime | None


class ApplicationSummaryResponse(BaseModel):
    id: UUID
    candidate_id: UUID
    candidate_name: str
    job_requirement_id: UUID
    job_title: str
    client_name: str
    pipeline_stage: PipelineStage
    status: ApplicationStatus
    has_parsed_resume: bool
    resume_version_count: int
    latest_resume_version_id: UUID | None = None
    latest_built_resume_status: ResumeVersionStatus | None = None
    latest_score: Decimal | None
    interview_status: InterviewStatus | None
    interview_overall_score: Decimal | None = None
    created_at: datetime
    updated_at: datetime


class ApplicationDetailResponse(BaseModel):
    id: UUID
    tenant_id: UUID
    candidate_id: UUID
    candidate: CandidateResponse
    job_requirement_id: UUID
    job_title: str
    client_name: str
    job_description: str | None
    job_experience_min_years: int | None = None
    job_experience_max_years: int | None = None
    pipeline_stage: PipelineStage
    status: ApplicationStatus
    recruiter_notes: str | None
    parsed_resume: ParsedResumeResponse | None
    resume_versions: list[ResumeVersionResponse]
    resume_scores: list[ResumeScoreResponse]
    screening_interview: ScreeningInterviewResponse | None
    created_at: datetime
    updated_at: datetime


class ApplicationListResponse(BaseModel):
    items: list[ApplicationSummaryResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class BuildResumeRequest(BaseModel):
    recruiter_notes: str | None = None
    template_id: UUID | None = None
    gap_strategy: str = Field(default="none", max_length=50)
    target_total_experience_years: float | None = Field(default=None, ge=0, le=50)
    summary: str | None = None
    skills: list[str] | None = None
    experience: list[dict[str, Any]] | None = None
    education: list[dict[str, Any]] | None = None
    training_entry: dict[str, Any] | None = None


class TimelineGapResponse(BaseModel):
    label: str
    from_date: str
    to_date: str
    months: int
    suggestion: str


class GapStrategyOption(BaseModel):
    id: str
    label: str
    description: str


class ResumeBuildContextResponse(BaseModel):
    job_experience_min_years: int | None
    job_experience_max_years: int | None
    candidate_total_experience_years: float
    experience_shortfall_years: float | None
    timeline_gaps: list[TimelineGapResponse]
    gap_strategy_options: list[GapStrategyOption]
    recommendations: list[str]
    default_summary: str | None
    default_skills: list[str]
    default_experience: list[dict[str, Any]]
    default_education: list[dict[str, Any]]
    templates: list[ResumeTemplateSummary] = []
    default_template_id: UUID | None = None


class UpdateResumeVersionRequest(BaseModel):
    content_json: dict[str, Any]


class RecruiterReviewRequest(BaseModel):
    decision: RecruiterReviewDecision
    notes: str | None = None


class ScheduleInterviewRequest(BaseModel):
    scheduled_at: datetime | None = None
    duration_minutes: int = Field(default=45, ge=15, le=180)
    difficulty: str = Field(default="medium", max_length=50)
    coding_required: bool = False
    behavioral: bool = True
    technical: bool = True
    language: str = Field(default="en", max_length=50)
    timezone: str = Field(default="UTC", max_length=100)


class InterviewAnswerRequest(BaseModel):
    answer: str = Field(..., min_length=1, max_length=10000)


class InterviewQuestionResponse(BaseModel):
    question: str
    interview_status: InterviewStatus
    question_number: int
    question_type: str = "mcq"
    options: list[dict[str, str]] = Field(default_factory=list)
    phase_label: str | None = None
    total_questions: int = 18
    is_complete: bool = False


class MessageResponse(BaseModel):
    message: str
